File: kiss_manga.js

var directory = {}
var manga = {}

function handleRequest(req) {
    api.note("JS, handling request!")
    var r_filter  = req.getFilter()
    var r_manga   = req.getManga()
    var r_chapter = req.getChapter()
    var r_page    = req.getPage()

    if (r_filter == "")   {
        return ["All", "Most Popular", "Latest Update", "Newest"]
    } else if (r_manga == "") {
        if (!directory[r_filter])
            finishDirectorySetup(r_filter)
        return directory[r_filter]
    } else if (r_chapter == "") {
        api.note("request for chapter list")
        if (!directory[r_filter])
            finishDirectorySetup(r_filter)
        if (!manga[r_manga]['chapters'])
            finishMangaSetup(r_manga)
        return [ manga[r_manga]['description'] ].concat(Object.keys(manga[r_manga]['chapters']))
    } else if (r_page == "")    {
        api.note("request for num pages")
        if (!directory[r_filter])
            finishDirectorySetup(r_filter)
        if (!manga[r_manga]['chapters'])
            finishMangaSetup(r_manga)
        if (!manga[r_manga]['chapters'][r_chapter]['pages'])
            finishChapterSetup(r_manga, r_chapter)
        return [ manga[r_manga]['chapters'][r_chapter]['pages'].length.toString() ]
    } else {
        api.note("request for page in " + r_manga + " : " + r_chapter + " : " + r_page)
        if (!directory[r_filter])
            finishDirectorySetup(r_filter)
        if (!manga[r_manga]['chapters'])
            finishMangaSetup(r_manga)
        if (!manga[r_manga]['chapters'][r_chapter]['pages'])
            finishChapterSetup(r_manga, r_chapter)
        return [ downloadCF(manga[r_manga]['chapters'][r_chapter]['pages'][r_page]) ]
    }
}

function finishDirectorySetup(r_filter) {
    api.note("request for manga list")
    var pageNo = 1
    var url = 'http://kissmanga.com/MangaList?page=' + pageNo
    if        (r_filter == "All") {
        // default yo
    } else if (r_filter == "Most Popular") {
        url = 'http://kissmanga.com/MangaList/MostPopular?page=' + pageNo
    } else if (r_filter =="Latest Update") {
        url = 'http://kissmanga.com/MangaList/LatestUpdate?page=' + pageNo
    } else if (r_filter =="Newest") {
        url = 'http://kissmanga.com/MangaList/Newest?page=' + pageNo
    }
    var path = downloadCF(url)
    var pageSource = api.readFile(path)

    var regex = /<\/div>'>[\s\S]*?<a href="\/Manga\/([\s\S]*?)">([\s\S]*?)<\/a>/g
    var match
    directory[r_filter] = []
    while (match = regex.exec(pageSource)) {
        manga[match[2]] = {title: match[2], url: match[1]}
        directory[r_filter].push(match[2])
    }
}

function finishMangaSetup(r_manga) {
    api.note("FINISHING MANGA SETUP FOR " + r_manga)
    var mangaURL = 'http://kissmanga.com/Manga' + '/' + manga[r_manga]['url'] + '?confirm=yes'
    var path = downloadCF(mangaURL)
    var pageSource = api.readFile(path)
    var descriptionRegex = /<span class="info">Summary:<\/span>[\s\S]*?<p[\s\S]*?>([\s\S]*?)<\/p>/g
    manga[r_manga]['description'] = descriptionRegex.exec(pageSource)[1]

    var chapterRegex = new RegExp('<a +href="/Manga/' + escapeRegExp(manga[r_manga]['url']) + '/([\\s\\S]*?)"[\\s\\S]*?>([\\s\\S]*?)</a>', 'g')
    var match
    manga[r_manga]['chapters'] = {}
    while (match = chapterRegex.exec(pageSource)) {
        var chapter = {title: match[2], url: match[1]}
        manga[r_manga]['chapters'][chapter.title] = chapter
    }
}

function finishChapterSetup(r_manga, r_chapter) {
    api.note("FINISHING CHAPTER SETUP FOR " + r_manga + ":" + r_chapter)
    var pagesURL = 'http://kissmanga.com/Manga/' + manga[r_manga]['url'] + '/' + manga[r_manga]['chapters'][r_chapter]['url']

    api.note('The Pages URL is: ' + pagesURL)
    var path = downloadCF(pagesURL)
    api.note('After download')
    var pageSource = api.readFile(path)

    //let's get with it
    //this won't stop us
    //they encode their links now, so we download all their crypto js
    //and execute it (over and over for every link right now...)
    path = downloadCF('http://kissmanga.com/Scripts/ca.js')
    var ca_js = api.readFile(path)
    path = downloadCF('http://kissmanga.com/Scripts/lo.js')
    var lo_js = api.readFile(path)
    var pageJS = ca_js + ';' + lo_js

    //var _0xa5a2 = ["\x37\x32\x6E\x6E\x61\x73\x64\x61\x73\x64\x39\x61\x73\x    64\x6E\x31\x32\x33"]; chko = _0xa5a2[0]; key = CryptoJS.SHA256(chko)
    //var _0x2c7e = ["\x6E\x61\x73\x64\x62\x61\x73\x64\x36\x31\x32\x62\x61\x    73\x64"]; chko = chko + _0x2c7e[0]; key = CryptoJS.SHA256(chko)

    var varkeyjs_regex = /(var _\w*? = \["[^"]*?"\]; *?chko = \w*? *?\+? *?_\w*?\[0\]; *?key = CryptoJS.SHA256\(chko\))/g
    var match
    var results = []
    while (match = varkeyjs_regex.exec(pageSource)) {
        api.note("Found key stuff " + match[1])
        pageJS += ";" + match[1]
    }
    pageJS += ';var message = "no message"; function alert(a) { message = a };'
    // eval the stuff once
    eval(pageJS)

    var regex = /lstImages\.push\((.*?)\);/g
    var match
    var results = []
    var numPages = 0
    manga[r_manga]['chapters'][r_chapter]['pages'] = []
    while (match = regex.exec(pageSource)) {
        var pageURL = match[1]
        api.note("evaling " + pageURL)
        var url = eval(pageURL)
        manga[r_manga]['chapters'][r_chapter]['pages'].push(url)
    }
    api.note("Pages found")
    api.note(manga[r_manga]['chapters'][r_chapter]['pages'])
}

File: kiss_manga.lua
pageNo = 1

function getMangaListTypes()
    titleList = { }
    titleList[0] = 'All'
    titleList[1] = 'Most Popular'
    titleList[2] = 'Latest Update'
    titleList[3] = 'Newest'
    titleList['numTypes'] = 4
    allChar = '#ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    i = 1
    while i < (string.len(allChar) + 1) do
        titleList[titleList['numTypes'] + i - 1] = string.sub(allChar, i, i)
        i = i + 1
    end
    titleList['numTypes'] = titleList['numTypes'] + i
    return titleList
end

function getMangaListPage(mangaListType)
    print('type is ' .. mangaListType)
    if mangaListType == 'All' then
        return getMangaList('http://kissmanga.com/MangaList?page=' .. pageNo)
    elseif mangaListType == 'Most Popular' then
        return getMangaList('http://kissmanga.com/MangaList/MostPopular?page=' .. pageNo)
    elseif mangaListType == 'Latest Update' then
        return getMangaList('http://kissmanga.com/MangaList/LatestUpdate?page=' .. pageNo)
    elseif mangaListType == 'Newest' then
        return getMangaList('http://kissmanga.com/MangaList/Newest?page=' .. pageNo)
    elseif mangaListType == '#' then
        return getMangaList('http://kissmanga.com/MangaList?c=0&page=' .. pageNo)
    end

    return getMangaList('http://kissmanga.com/MangaList?c=' .. mangaListType .. '&page=' .. pageNo)
end

function getMangaList(url)
   print('About to getMangaList!')
   print(apiObj)
   path = download_cf(url)
   pageSource = apiObj:readFile(path)
   apiObj:note('LuaScript downloaded (for manga): ' .. path)
   daList = {}
   regex = '</div>\'>.-<a href="/Manga/(.-)">(.-)</a>'
   apiObj:note('Manga List Regex: ' .. regex)
   beginning, ending, mangaURL, mangaTitle = string.find(pageSource, regex)
   index = 0
   while ending do
       daList[index] = {title = mangaTitle, url = mangaURL}
       beginning, ending, mangaURL, mangaTitle = string.find(pageSource, regex, ending+1)
       index = index + 1
   end
   daList['numManga'] = index
   return daList
end


function initManga(manga)
   mangaURL = 'http://kissmanga.com/Manga' .. '/' .. manga['url'] .. '?confirm=yes'
   apiObj:note('Manga Path: ' .. mangaURL)
   path = download_cf(mangaURL)
   pageSource = apiObj:readFile(path)
   apiObj:note('LuaScript downloaded (for chapter): ' .. path)

   -- Set up manga description and other nicities
   descriptionRegex = '<span class="info">Summary:</span>.-<p.->(.-)</p>'
   _, _, mangaDescription = string.find(pageSource, descriptionRegex)
   manga['description'] = mangaDescription

   daList = {}
   --regex = '<a +href="/Manga/' .. escapeRegexStr(manga['url']) .. '/(.-)?id.-".->(.-)</a>'
   regex = '<a +href="/Manga/' .. escapeRegexStr(manga['url']) .. '/(.-)".->(.-)</a>'
   apiObj:note('Chapter List Regex: ' .. regex)
   beginning, ending, chapterURL, chapterTitle = string.find(pageSource, regex)
   index = 0
   while ending do
       daList[index] = {title = chapterTitle, url = chapterURL, chapterSetUp = false}
       beginning, ending, chapterURL, chapterTitle = string.find(pageSource, regex, ending+1)
       index = index + 1
   end
   daList['numChapters'] = index
   manga['chapter_list'] = daList
end

function getMangaChapterList(manga)
    return manga['chapter_list']
end


function getMangaChapterNumPages(manga, chapter)
   if not chapter['chapterSetUp'] then
       setUpChapter(manga, chapter)
   end
   return chapter['pageList']['numPages']
end


function getMangaChapterPage(manga, chapter, page)
   if not chapter['chapterSetUp'] then
       setUpChapter(manga, chapter)
   end
   apiObj:note('lookie')
   apiObj:note('there are # pages')
   apiObj:note(chapter['pageList']['numPages'])
   apiObj:note(chapter['pageList'])
   apiObj:note(chapter['pageList'][page])
   apiObj:note(chapter['pageList'][page]['url'])
   return download_cf(chapter['pageList'][page]['url'])
end


function setUpChapter(manga, chapter)

       pageURL = 'http://kissmanga.com/Manga' .. '/' .. manga['url'] .. '/' .. chapter['url']

       apiObj:note('The Page URL is: ' .. pageURL)
       path = download_cf(pageURL)
       apiObj:note('After download')
       pageSource = apiObj:readFile(path)


       --let's get with it
       --this won't stop us
       --they encode their links now, so we download all their crypto js
       --and execute it (over and over for every link right now...)
       path = download_cf('http://kissmanga.com/Scripts/ca.js')
       ca_js = apiObj:readFile(path)
       path = download_cf('http://kissmanga.com/Scripts/lo.js')
       lo_js = apiObj:readFile(path)
       pageJS = ca_js .. ';' .. lo_js

--var _0xa5a2 = ["\x37\x32\x6E\x6E\x61\x73\x64\x61\x73\x64\x39\x61\x73\x    64\x6E\x31\x32\x33"]; chko = _0xa5a2[0]; key = CryptoJS.SHA256(chko)
--var _0x2c7e = ["\x6E\x61\x73\x64\x62\x61\x73\x64\x36\x31\x32\x62\x61\x    73\x64"]; chko = chko + _0x2c7e[0]; key = CryptoJS.SHA256(chko)

       keyjs_regex = '(var _[%a%d]- = %["[^"]-"%]; -chko = %a- -+? -_[%a%d]-%[0%]; -key = CryptoJS.SHA256%(chko%))'
       beginning, ending, keyjs = string.find(pageSource, keyjs_regex)
       while ending do
           pageJS = pageJS .. ';' .. keyjs
           beginning, ending, keyjs = string.find(pageSource, keyjs_regex, ending+1)
       end
       pageJS = pageJS .. ';var message = "no message"; function alert(a) { message = a };'


       regex = 'lstImages%.push%((.-)%);'
       apiObj:note('Page List Regex: ' .. regex)
       beginning, ending, pageURL = string.find(pageSource, regex)
       index = 0
       daList = {}
       while ending do
           daList[index] = {url = apiObj:doDaJS(pageJS .. pageURL)}
           apiObj:note("orig url is")
           apiObj:note(pageURL)
           apiObj:note("decoded url is")
           apiObj:note(daList[index]['url'])
           beginning, ending, pageURL = string.find(pageSource, regex, ending+1)
           index = index + 1
       end
       daList['numPages'] = index
       chapter['pageList'] = daList
       chapter['chapterSetUp'] = true
end

s_manga_list = {}
s_chapter_list = {}
s_manga = {}
s_chapter = {}

function handleRequest(req)
    print("Lua, handling request!")
    r_filter = req:getFilter()
    r_manga = req:getManga()
    r_chapter = req:getChapter()
    r_page = req:getPage()
    ret = {}

    if r_manga == "" then
        -- This is a request for manga listing
       mangaList = getMangaListPage(r_filter)
       print("Request got manga list")
       index = 0
       for k,v in ipairs(mangaList) do
           manga_title = v['title']
           manga_url = v['url']
           s_manga_list[manga_title] = manga_url
           ret[index] = manga_title
           index = index + 1
       end
       return ret
    end

    if r_chapter == "" then
        -- This is a request for chapter listing
        print("Got a request for a chapter listing")
        s_manga = { }
        s_manga['url'] = s_manga_list[r_manga]
        initManga(s_manga)
        chap_list = s_manga['chapter_list']
        ret[0] = s_manga['description']
        index = 1
        for k,v in pairs(chap_list) do
            if k ~= 'numChapters' then
                ch_title = v['title']
                ch_url = v['url']
                ret[index] = ch_title
                s_chapter_list[ch_title] = ch_url
                index = index + 1
            end
        end
        return ret
    end

    if r_page == "" then
        -- Requesting the number of pages for the chapter
        s_chapter = {}
        s_chapter['url'] = s_chapter_list[r_chapter]
        s_chapter['chapterSetUp'] = false
        tmp = getMangaChapterNumPages(s_manga,s_chapter)
        print("Page Number Request")
        print(tmp)
        ret[0] = tostring(tmp)
        ret[1] = tostring(tmp)
        ret['numPages'] = tostring(tmp)
        return ret
    end

    -- We have a page request
    print("Requesting page")
    print(r_page)
    s_chapter['url'] = s_chapter_list[r_chapter]
    page_url = getMangaChapterPage(s_manga, s_chapter, tonumber(r_page))
    print("page url is...")
    print(page_url)
    ret[0] = page_url
    ret[1] = page_url
    return ret
end
